File: csharp/Contexel/Stages.cs
using System.Globalization;
using System.Text;
using System.Text.RegularExpressions;

namespace Contexel;

using Rec = Dictionary<string, object?>;
using Records = List<Dictionary<string, object?>>;

/// <summary>
/// Deterministic context-economy stages — parity port of contexel's Python
/// stages module. Every stage takes and returns records, never mutates its
/// input, and is auto-traced when a trace is active. Behavior matches the
/// Python canon on the shared golden vectors (parity/vectors.json);
/// boundaries are documented in ADR-002.
/// </summary>
public static class Stages
{
    /// <summary>A bare string names ONE field, never its characters.</summary>
    private static List<string> FieldList(object fields) => fields switch
    {
        string s => [s],
        IEnumerable<string> seq => seq.ToList(),
        _ => throw new ArgumentException("fields must be a string or a sequence of strings"),
    };

    /* ---------------------------- fingerprints --------------------------- */

    /// <summary>Type-qualified fingerprint (1, 1.0, "1", and true all
    /// differ — C# preserves int/float, exceeding the TS port). Containers
    /// are type-tagged so {} and [] never collide (canon fix, 0.1.16).</summary>
    internal static string TypedKey(object? value) => value switch
    {
        null => "NoneType:None",
        bool b => "bool:" + (b ? "True" : "False"),
        long l => "int:" + l.ToString(CultureInfo.InvariantCulture),
        int i => "int:" + i.ToString(CultureInfo.InvariantCulture),
        // -0.0 == 0.0 in Python (equal dict keys): normalize before keying
        double d => "float:" + PyCompat.PyNumber(d == 0 ? 0.0 : d),
        string s => "str:" + Tokens.PythonJson(s),
        System.Collections.IDictionary dict =>
            "dict:{" + string.Join(",", dict.Keys.Cast<object>()
                .Select(k => Tokens.PythonJson((string)k) + ":" + TypedKey(dict[k]))
                .OrderBy(x => x, StringComparer.Ordinal)) + "}",
        System.Collections.IEnumerable seq =>
            "list:[" + string.Join(",", seq.Cast<object?>().Select(TypedKey)) + "]",
        _ => value.GetType().Name + ":" +
             Convert.ToString(value, System.Globalization.CultureInfo.InvariantCulture),
    };

    /* ------------------------------- stages ------------------------------ */

    public static Records Select(Records records, object fields) =>
        Trace.Traced("select", records, () =>
        {
            var keep = FieldList(fields);
            return records.Select(r =>
            {
                var output = new Rec();
                foreach (var k in keep)
                    if (r.TryGetValue(k, out var v)) output[k] = v;
                return output;
            }).ToList();
        });

    public static Records Dedupe(Records records, object? key = null) =>
        Trace.Traced("dedupe", records, () =>
        {
            List<string>? keys = key is null ? null : FieldList(key);
            var seen = new HashSet<string>();
            var output = new Records();
            foreach (var r in records)
            {
                object? basis = keys is null
                    ? r
                    : keys.Select(k => r.TryGetValue(k, out var v) ? v : null).ToList<object?>();
                string fp = TypedKey(basis);
                if (seen.Add(fp)) output.Add(r);
            }
            return output;
        });

    public static Records Allowlist(Records records, string field, IEnumerable<object?> allowed) =>
        Trace.Traced("allowlist", records, () =>
        {
            // Python builds set(allowed) up front: unhashable entries raise
            var allowedKeys = new HashSet<string>(allowed.Select(PyCompat.PyEqKey));
            return records.Where(r =>
            {
                // r.get(field): missing reads as None, which may be allowed;
                // unhashable record values fail closed
                var v = r.TryGetValue(field, out var x) ? x : null;
                try { return allowedKeys.Contains(PyCompat.PyEqKey(v)); }
                catch (ArgumentException) { return false; }
            }).ToList();
        });

    internal static readonly string[] InjectionPatterns =
    [
        @"ignore\s+(?:all\s+|any\s+)?(?:previous|prior|above|earlier)\s+(?:instructions|messages|prompts|context)",
        @"disregard\s+(?:all\s+|the\s+)?(?:previous|prior|above|earlier)",
        @"you\s+are\s+now\s+",
        @"new\s+instructions\s*:",
        @"system\s+prompt",
        @"</?\s*(?:system|assistant|instructions?)\s*>",
    ];

    public static Records Quarantine(Records records, object? fields = null,
                                     object? patterns = null, string action = "drop",
                                     string into = "quarantined",
                                     bool replacePatterns = false) =>
        Trace.Traced("quarantine", records, () =>
        {
            if (action is not ("drop" or "flag"))
                throw new ArgumentException($"action must be 'drop' or 'flag', got '{action}'");
            var fieldNames = FieldList(fields ?? new[] { "snippet" });
            // custom patterns EXTEND the built-ins: domain markers must never
            // silently disable "ignore all previous instructions" detection;
            // replacePatterns is the explicit opt-out
            List<string> patternList;
            if (patterns is null)
            {
                if (replacePatterns)
                    throw new ArgumentException(
                        "replacePatterns without patterns is contradictory; pass the replacement patterns");
                patternList = InjectionPatterns.ToList();
            }
            else
            {
                var userPatterns = FieldList(patterns);
                if (userPatterns.Any(string.IsNullOrEmpty))
                    throw new ArgumentException("empty pattern fragments match everything; remove them");
                patternList = replacePatterns
                    ? userPatterns
                    : InjectionPatterns.Concat(userPatterns).ToList();
            }
            if (patternList.Count == 0)
                throw new ArgumentException(
                    "replacePatterns with no patterns would disable the control entirely; pass at least one pattern");
            var matcher = new Regex(string.Join("|", patternList),
                                    RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
            var output = new Records();
            foreach (var r in records)
            {
                // Python str(r.get(f, "")): None -> "None", containers via
                // repr — a payload nested in a dict or list still matches
                string text = string.Join(" ", fieldNames.Select(f =>
                    r.TryGetValue(f, out var v) ? PyCompat.PyStr(v) : ""));
                bool hit = matcher.IsMatch(text);
                if (action == "flag")
                {
                    var flagged = new Rec(r) { [into] = hit };
                    output.Add(flagged);
                }
                else if (!hit) output.Add(r);
            }
            return output;
        });

    // Python's [^\W_] on ASCII; exotic Unicode word classes are a margin
    private static readonly Regex Word =
        new(@"[\p{L}\p{N}]+", RegexOptions.CultureInvariant);

    public static Records Rescore(Records records, object query, object? fields = null,
                                  string into = "score", bool proximity = true,
                                  string match = "word") =>
        Trace.Traced("rescore", records, () =>
        {
            var fieldNames = FieldList(fields ?? new[] { "snippet" });
            var rawTerms = (query switch
            {
                string q => Word.Matches(q.ToLowerInvariant()).Select(m => m.Value),
                IEnumerable<string> qs => qs.SelectMany(p =>
                    Word.Matches(p.ToLowerInvariant()).Select(m => m.Value)),
                _ => throw new ArgumentException("query must be a string or a sequence of strings"),
            }).ToList();
            var terms = rawTerms.Distinct().ToList();
            if (terms.Count == 0 || records.Count == 0)
                return records.Select(r => new Rec(r) { [into] = 0.0 }).ToList();

            var texts = records.Select(r => string.Join(" ", fieldNames.Select(f =>
                r.TryGetValue(f, out var v) ? PyCompat.PyStr(v) : "")).ToLowerInvariant()).ToList();

            Regex? finder = null;
            if (match == "word")
            {
                string alt = string.Join("|", terms
                    .OrderByDescending(t => t.Length)
                    .Select(Regex.Escape));
                // boundary guards: underscore counts as a boundary
                finder = new Regex($@"(?<![\p{{L}}\p{{N}}])(?:{alt})(?![\p{{L}}\p{{N}}])",
                                   RegexOptions.CultureInvariant);
            }

            int n = records.Count;
            var tfs = new List<Dictionary<string, int>>();
            var firsts = new List<Dictionary<string, int>>();
            var df = terms.ToDictionary(t => t, _ => 0);
            foreach (var text in texts)
            {
                // Python positions are code points; .NET Match.Index is UTF-16
                var toCp = PyCompat.CodePointIndexer(text);
                var row = new Dictionary<string, int>();
                var first = new Dictionary<string, int>();
                if (finder is not null)
                {
                    foreach (Match m in finder.Matches(text))
                    {
                        string t = m.Value;
                        row[t] = row.GetValueOrDefault(t) + 1;
                        if (!first.ContainsKey(t)) first[t] = toCp(m.Index);
                    }
                }
                else
                {
                    foreach (var t in terms)
                    {
                        int idx = text.IndexOf(t, StringComparison.Ordinal);
                        int cnt = 0, firstIdx = idx;
                        while (idx != -1)
                        {
                            cnt++;
                            // Python str.count is NON-overlapping: advance
                            // past the whole match ("aaaa".count("aa") == 2)
                            idx = text.IndexOf(t, idx + t.Length, StringComparison.Ordinal);
                        }
                        if (cnt > 0) { row[t] = cnt; first[t] = toCp(firstIdx); }
                    }
                }
                foreach (var t in row.Keys) df[t] += 1;
                tfs.Add(row);
                firsts.Add(first);
            }
            var idf = terms.ToDictionary(t => t,
                t => Math.Log((n - df[t] + 0.5) / (df[t] + 0.5) + 1.0));

            const double k1 = 1.5;
            const int window = 80;
            return records.Select((r, i) =>
            {
                var row = tfs[i];
                var first = firsts[i];
                double score = 0;
                foreach (var (t, tf) in row) score += idf[t] * (tf / (tf + k1));
                if (proximity && row.Count > 1)
                {
                    for (int j = 0; j + 1 < terms.Count; j++)
                    {
                        string t1 = terms[j], t2 = terms[j + 1];
                        if (first.TryGetValue(t1, out int p1) && first.TryGetValue(t2, out int p2))
                        {
                            int gap = p2 - p1;
                            if (gap > 0 && gap <= window) score += (idf[t1] + idf[t2]) / 4;
                        }
                    }
                }
                // Python's score starts as int 0 and becomes float only when
                // a term contributes: a no-match record carries int 0, which
                // serializes "0", not "0.0" — C# preserves that distinction
                // explicit boxing: the ternary would otherwise unify long
                // and double to double, silently turning 0 into 0.0
                object scoreValue = row.Count == 0 ? 0L : (object)PyCompat.PyRound(score, 6);
                return new Rec(r) { [into] = scoreValue };
            }).ToList();
        });

    public static Records Rank(Records records, string by, bool desc = true) =>
        Trace.Traced("rank", records, () =>
        {
            var present = records.Where(r => r.ContainsKey(by)).ToList();
            var missing = records.Where(r => !r.ContainsKey(by)).ToList();
            return StableSort(present, r => r[by], desc).Concat(missing).ToList();
        });

    public static Records Rank(Records records, Func<Rec, object?> by, bool desc = true) =>
        Trace.Traced("rank", records, () => StableSort(records, by, desc));

    // LINQ OrderBy is stable (List.Sort is NOT); a directional comparator
    // with a stable sort keeps ties in original order like Python
    // sorted(key=..., reverse=...). OrderBy wraps comparer exceptions in
    // InvalidOperationException — unwrap so incomparable values raise
    // ArgumentException like Python's TypeError.
    private static Records StableSort(Records records, Func<Rec, object?> keyOf, bool desc)
    {
        int dir = desc ? -1 : 1;
        var comparer = Comparer<object?>.Create((a, b) => dir * PyCompat.PyCompare(a, b));
        try
        {
            return records.OrderBy(keyOf, comparer).ToList();
        }
        catch (InvalidOperationException e) when (e.InnerException is ArgumentException inner)
        {
            throw new ArgumentException(inner.Message, inner);
        }
    }

    private static string TruncateText(string text, int maxTokens,
                                       Func<string, int>? tokenizer, string suffix)
    {
        if (Tokens.Count(text, tokenizer) <= maxTokens) return text;
        int budget = Math.Max(1, maxTokens - Tokens.Count(suffix, tokenizer));
        if (Tokens.IsHeuristic(tokenizer))
        {
            // ceil(len/4) is invertible: longest fitting prefix = 4*budget
            // code points (Python slices by code points, so must we)
            return PyCompat.PyRstrip(PyCompat.CodePointSlice(text, 4 * budget)) + suffix;
        }
        int lo = 0, hi = PyCompat.CodePointLength(text);
        while (lo < hi)
        {
            int mid = (lo + hi + 1) >> 1;
            if (Tokens.Count(PyCompat.CodePointSlice(text, mid), tokenizer) <= budget) lo = mid;
            else hi = mid - 1;
        }
        return PyCompat.PyRstrip(PyCompat.CodePointSlice(text, lo)) + suffix;
    }

    public static Records TruncateField(Records records, string field, int maxTokens,
                                        Func<string, int>? tokenizer = null,
                                        string suffix = "…") =>
        Trace.Traced("truncate_field", records, () =>
            records.Select(r =>
            {
                if (!r.TryGetValue(field, out var v) || v is not string s) return r;
                return new Rec(r) { [field] = TruncateText(s, maxTokens, tokenizer, suffix) };
            }).ToList());

    public static Records TrimToBudget(Records records, int maxTokens,
                                       Func<string, int>? tokenizer = null,
                                       int minRecords = 0) =>
        Trace.Traced("trim_to_budget", records, () =>
        {
            // minRecords guards the silent-failure edge: a too-small budget
            // would return [], indistinguishable from "nothing matched"
            var output = new Records();
            int used = 0;
            foreach (var r in records)
            {
                int cost = Tokens.Count(r, tokenizer) + 2; // list framing
                if (used + cost > maxTokens && output.Count >= minRecords) break;
                output.Add(r);
                used += cost;
            }
            return output;
        });

    /// <summary>Combiner (not auto-traced, matching Python): normalize and
    /// concatenate record lists from different tools; schema maps unified
    /// name -> source name or candidate list (first match wins).</summary>
    public static Records Merge(IEnumerable<Records> sources,
                               Dictionary<string, object>? schema = null,
                               string? dedupeKey = null)
    {
        var combined = new Records();
        foreach (var src in sources)
        {
            foreach (var r in src)
            {
                if (schema is null) { combined.Add(r); continue; }
                var mapped = new Rec();
                foreach (var (unified, candidates) in schema)
                {
                    List<string> names = candidates switch
                    {
                        string one => [one],
                        IEnumerable<string> many => many.ToList(),
                        _ => throw new ArgumentException("schema values must be a string or a sequence of strings"),
                    };
                    foreach (var name in names)
                    {
                        if (r.TryGetValue(name, out var v)) { mapped[unified] = v; break; }
                    }
                }
                combined.Add(mapped);
            }
        }
        return dedupeKey is null ? combined : Dedupe(combined, dedupeKey);
    }
}

File: ts/src/stages.ts
/**
 * Deterministic context-economy stages — parity port of contexel's Python
 * `stages` module. Every stage is `(records, params) => records`, never
 * mutates its input, and is auto-traced when a trace is active. Behavior
 * matches the Python implementation on the shared golden vectors
 * (parity/vectors.json); boundaries are documented in ADR-001.
 */
import * as tokens from "./tokens.js";
import { traced } from "./trace.js";
import {
  codePointIndexer,
  pyCompare,
  pyEqKey,
  pyRound,
  pyRstrip,
  pyStr,
} from "./py.js";

export type Rec = Record<string, unknown>;
export type Records = Rec[];

/** A bare string means ONE field, not its characters (parity with Python). */
function fieldList(fields: string | string[]): string[] {
  return typeof fields === "string" ? [fields] : fields;
}

/** Own-property write that cannot hit setters or the "__proto__" trap —
 * a key named "__proto__" becomes an ordinary data property, as in Python. */
function defineField(out: Rec, key: string, value: unknown): void {
  Object.defineProperty(out, key, {
    value,
    writable: true,
    enumerable: true,
    configurable: true,
  });
}

/* ------------------------------ fingerprints ----------------------------- */

/** Python-mapped type tag: int/float distinction for JSON-sourced numbers. */
function typedKey(value: unknown): string {
  if (value === null || value === undefined) return "NoneType:None";
  if (typeof value === "boolean") return `bool:${value}`;
  if (typeof value === "number") {
    return Number.isInteger(value) ? `int:${value}` : `float:${value}`;
  }
  if (typeof value === "string") return `str:${JSON.stringify(value)}`;
  if (Array.isArray(value)) return "list:[" + value.map(typedKey).join(",") + "]";
  if (value instanceof Set) {
    return "set:(" + [...value].map(typedKey).sort().join(",") + ")";
  }
  if (typeof value === "object") {
    const entries = Object.entries(value as Rec)
      .map(([k, v]) => `${JSON.stringify(k)}:${typedKey(v)}`)
      .sort();
    return "dict:{" + entries.join(",") + "}";
  }
  return `${typeof value}:${String(value)}`;
}

/* -------------------------------- stages --------------------------------- */

export const select = traced(
  "select",
  (records: Records, params: { fields: string | string[] }): Records => {
    const keep = fieldList(params.fields);
    return records.map((r) => {
      const out: Rec = {};
      for (const k of keep) if (Object.hasOwn(r, k)) defineField(out, k, r[k]);
      return out;
    });
  }
);

export const dedupe = traced(
  "dedupe",
  (records: Records, params: { key?: string | string[] | null } = {}): Records => {
    const key = params.key ?? null;
    const keys = key === null ? null : Array.isArray(key) ? key : [key];
    const seen = new Set<string>();
    const out: Records = [];
    for (const r of records) {
      const basis =
        keys === null ? r : keys.map((k) => (Object.hasOwn(r, k) ? r[k] : null));
      const fp = typedKey(basis);
      if (seen.has(fp)) continue;
      seen.add(fp);
      out.push(r);
    }
    return out;
  }
);

export const allowlist = traced(
  "allowlist",
  (records: Records, params: { field: string; allowed: unknown[] | Set<unknown> }): Records => {
    // Python builds set(allowed) up front: an unhashable entry raises.
    const allowed = new Set<string>();
    for (const a of params.allowed) allowed.add(pyEqKey(a));
    return records.filter((r) => {
      // Python r.get(field): a missing field reads as None, and None can
      // legitimately be allowed; unhashable values fail closed.
      const v = Object.hasOwn(r, params.field) ? r[params.field] : null;
      try {
        return allowed.has(pyEqKey(v));
      } catch {
        return false;
      }
    });
  }
);

const INJECTION_PATTERNS = [
  "ignore\\s+(?:all\\s+|any\\s+)?(?:previous|prior|above|earlier)\\s+(?:instructions|messages|prompts|context)",
  "disregard\\s+(?:all\\s+|the\\s+)?(?:previous|prior|above|earlier)",
  "you\\s+are\\s+now\\s+",
  "new\\s+instructions\\s*:",
  "system\\s+prompt",
  "</?\\s*(?:system|assistant|instructions?)\\s*>",
];

export const quarantine = traced(
  "quarantine",
  (
    records: Records,
    params: {
      fields?: string | string[];
      patterns?: string[];
      action?: "drop" | "flag";
      into?: string;
    } = {}
  ): Records => {
    const fields = fieldList(params.fields ?? ["snippet"]);
    const action = params.action ?? "drop";
    const into = params.into ?? "quarantined";
    if (action !== "drop" && action !== "flag") {
      throw new Error(`action must be 'drop' or 'flag', got ${String(action)}`);
    }
    const matcher = new RegExp((params.patterns ?? INJECTION_PATTERNS).join("|"), "i");
    const out: Records = [];
    for (const r of records) {
      // Python str(r.get(f, "")): None -> "None", nested containers via
      // repr — a payload inside a dict or list still hits the matcher.
      const text = fields
        .map((f) => (Object.hasOwn(r, f) ? pyStr(r[f]) : ""))
        .join(" ");
      const hit = matcher.test(text);
      if (action === "flag") out.push({ ...r, [into]: hit });
      else if (!hit) out.push(r);
    }
    return out;
  }
);

/* Word matching: identical to Python's [^\W_] on ASCII (all parity
   vectors); exotic Unicode word classes may differ at the margin. */
const WORD = /[\p{L}\p{N}]+/gu;

export const rescore = traced(
  "rescore",
  (
    records: Records,
    params: {
      query: string | string[];
      fields?: string | string[];
      into?: string;
      proximity?: boolean;
      match?: "word" | "substring";
    }
  ): Records => {
    const fields = fieldList(params.fields ?? ["snippet"]);
    const into = params.into ?? "score";
    const proximity = params.proximity ?? true;
    const match = params.match ?? "word";

    const rawTerms =
      typeof params.query === "string"
        ? params.query.toLowerCase().match(WORD) ?? []
        : params.query.flatMap((p) => String(p).toLowerCase().match(WORD) ?? []);
    const terms = [...new Set(rawTerms)];
    if (!terms.length || !records.length) {
      return records.map((r) => ({ ...r, [into]: 0.0 }));
    }

    const texts = records.map((r) =>
      fields
        .map((f) => (Object.hasOwn(r, f) ? pyStr(r[f]) : ""))
        .join(" ")
        .toLowerCase()
    );

    let finder: RegExp | null = null;
    if (match === "word") {
      const alt = [...terms]
        .sort((a, b) => b.length - a.length)
        .map((t) => t.replace(/[.*+?^${}()|[\]\\]/g, "\\$&"))
        .join("|");
      // boundary guards: underscore counts as a boundary (snake_case splits)
      finder = new RegExp(`(?<![\\p{L}\\p{N}])(?:${alt})(?![\\p{L}\\p{N}])`, "gu");
    }

    const n = records.length;
    const tfs: Array<Map<string, number>> = [];
    const firsts: Array<Map<string, number>> = [];
    const df = new Map<string, number>(terms.map((t) => [t, 0]));
    for (const text of texts) {
      // Python positions are code-point indices; JS regex/indexOf yield
      // UTF-16 units — convert so the proximity window means the same thing.
      const toCp = codePointIndexer(text);
      const row = new Map<string, number>();
      const first = new Map<string, number>();
      if (finder) {
        finder.lastIndex = 0;
        for (const m of text.matchAll(finder)) {
          const t = m[0];
          row.set(t, (row.get(t) ?? 0) + 1);
          if (!first.has(t)) first.set(t, toCp(m.index ?? 0));
        }
      } else {
        for (const t of terms) {
          let idx = text.indexOf(t);
          let cnt = 0;
          const firstIdx = idx;
          while (idx !== -1) {
            cnt++;
            idx = text.indexOf(t, idx + 1);
          }
          if (cnt) {
            row.set(t, cnt);
            first.set(t, toCp(firstIdx));
          }
        }
      }
      for (const t of row.keys()) df.set(t, (df.get(t) ?? 0) + 1);
      tfs.push(row);
      firsts.push(first);
    }
    const idf = new Map<string, number>(
      terms.map((t) => [t, Math.log((n - (df.get(t) ?? 0) + 0.5) / ((df.get(t) ?? 0) + 0.5) + 1.0)])
    );

    const k1 = 1.5;
    const window = 80;
    return records.map((r, i) => {
      const row = tfs[i];
      const first = firsts[i];
      let score = 0;
      for (const [t, tf] of row) score += (idf.get(t) ?? 0) * (tf / (tf + k1));
      if (proximity && row.size > 1) {
        for (let j = 0; j + 1 < terms.length; j++) {
          const t1 = terms[j];
          const t2 = terms[j + 1];
          if (first.has(t1) && first.has(t2)) {
            const gap = (first.get(t2) as number) - (first.get(t1) as number);
            if (gap > 0 && gap <= window) {
              score += ((idf.get(t1) ?? 0) + (idf.get(t2) ?? 0)) / 4;
            }
          }
        }
      }
      return { ...r, [into]: pyRound(score, 6) };
    });
  }
);

export const rank = traced(
  "rank",
  (
    records: Records,
    params: { by: string | ((r: Rec) => unknown); desc?: boolean }
  ): Records => {
    const desc = params.desc ?? true;
    const by = params.by;
    const dir = desc ? -1 : 1;
    // decorate-sort-undecorate like Python sorted(key=...): the key is
    // computed once per record, the sort is stable, and reverse=desc keeps
    // ties in original order (never reverse a stable ascending sort).
    const sortKeyed = (rs: Records, keyOf: (r: Rec) => unknown): Records => {
      const keyed = rs.map((r) => [keyOf(r), r] as const);
      keyed.sort((p, q) => dir * pyCompare(p[0], q[0]));
      return keyed.map((p) => p[1]);
    };
    if (typeof by === "function") return sortKeyed([...records], by);
    const present = records.filter((r) => Object.hasOwn(r, by));
    const missing = records.filter((r) => !Object.hasOwn(r, by));
    return [...sortKeyed(present, (r) => r[by]), ...missing];
  }
);

function truncateText(
  text: string,
  maxTokens: number,
  tokenizer: tokens.Tokenizer | null,
  suffix: string
): string {
  if (tokens.count(text, tokenizer) <= maxTokens) return text;
  const budget = Math.max(1, maxTokens - tokens.count(suffix, tokenizer));
  if (tokens.isHeuristic(tokenizer)) {
    // ceil(len/4) is invertible: longest prefix within budget = 4*budget
    // code points (Python slices by code points, so must we).
    const cps = [...text];
    return pyRstrip(cps.slice(0, 4 * budget).join("")) + suffix;
  }
  const cps = [...text];
  let lo = 0;
  let hi = cps.length;
  while (lo < hi) {
    const mid = (lo + hi + 1) >> 1;
    if (tokens.count(cps.slice(0, mid).join(""), tokenizer) <= budget) lo = mid;
    else hi = mid - 1;
  }
  return pyRstrip(cps.slice(0, lo).join("")) + suffix;
}

export const truncateField = traced(
  "truncate_field",
  (
    records: Records,
    params: {
      field: string;
      maxTokens: number;
      tokenizer?: tokens.Tokenizer | null;
      suffix?: string;
    }
  ): Records => {
    const suffix = params.suffix ?? "…";
    return records.map((r) => {
      const value = r[params.field];
      if (typeof value !== "string") return r;
      return {
        ...r,
        [params.field]: truncateText(value, params.maxTokens, params.tokenizer ?? null, suffix),
      };
    });
  }
);

export const trimToBudget = traced(
  "trim_to_budget",
  (
    records: Records,
    params: { maxTokens: number; tokenizer?: tokens.Tokenizer | null; minRecords?: number }
  ): Records => {
    // minRecords guards the silent-failure edge: a too-small budget would
    // return [], indistinguishable from "nothing matched" — keep the first
    // (best-ranked) records anyway; the overrun shows in the audit.
    const minRecords = params.minRecords ?? 0;
    const out: Records = [];
    let used = 0;
    for (const r of records) {
      const cost = tokens.count(r, params.tokenizer ?? null) + 2; // list framing
      if (used + cost > params.maxTokens && out.length >= minRecords) break;
      out.push(r);
      used += cost;
    }
    return out;
  }
);

/** Combiner (not auto-traced, matching Python): normalize and concatenate
 * record lists from different tools; `schema` maps unified name -> source
 * name or candidate list (first match wins). */
export function merge(
  sources: Records[],
  options: {
    schema?: Record<string, string | string[]>;
    dedupeKey?: string | null;
  } = {}
): Records {
  const combined: Records = [];
  for (const src of sources) {
    for (const r of src) {
      if (!options.schema) {
        combined.push(r);
        continue;
      }
      const mapped: Rec = {};
      for (const [unified, candidates] of Object.entries(options.schema)) {
        const names = typeof candidates === "string" ? [candidates] : candidates;
        for (const name of names) {
          if (Object.hasOwn(r, name)) {
            defineField(mapped, unified, r[name]);
            break;
          }
        }
      }
      combined.push(mapped);
    }
  }
  if (options.dedupeKey != null) {
    return dedupe(combined, { key: options.dedupeKey });
  }
  return combined;
}
